build_runner/lib/src/package_builder/package_builder.dart

```diff
@@ -0,0 +1,15 @@
+import 'package:build/build.dart';
+
+abstract class PackageBuilder extends Builder {
```

Comment:
If only we had type unions we wouldn't need the `extends` here...

I wonder if we'd be better off with a hierarchy like:

- `BuildAction`
- `AssetBuildAction extends BuildAction` - basically the implementation we have now and the default ctor for `BuildAction` is a factory that gives this.
- `PackageBuildAction extends BuildAction` - a new implementation that has a `PackageBuilder` instead of a `Builder`.  Then we shift the `is` check and have a more distinct separation between what can or can't be run with `package:build`  code.

Thoughts?

Reply:
I think that makes sense, I can work on updating it tomorrow. We can also remove the `inputs` field from the base class so its only on `AssetBuildAction`...which would be a little bit awkward since the `BuildAction` default constructor would be taking an arg for a field that doesn't exist on the interface, but I think it would be fine.

Reply:
Yeah I think I've seen cases where the default constructor instantiates a subclass and passes through arguments the parent doesn't actually use. In practice I don't think it feels very awkward, and we can make a named constructor like `new BuildAction.package()` for the other subclass.

Reply:
Ok, done.

I did not add a redirecting factory constructor on `BuildAction` for `PackageBuildAction`, but I could if you want. I don't think that is any cleaner than just having people go through the regular `PackageBuildAction` constructor though.

Reply:
This still has the `extends` did you find out that it isn't possible to remove that?

I was hoping for `BuildAction` to have no `builder` field, for `AssetBuildAction` to have a `builder` field of type `Builder`, and for `PackageBuildAction` to have a `builder` (or another name) field of type `PackageBuilder`, so that once we do the `is` check on the action type we get a concrete type for the builder we use.

Reply:
oh - I didn't make it so that PackageBuilder doesn't extend builder but I did make `PackageBuildAction` have a `PackageBuilder`, so you don't need any casts after checking for the type of the `BuildAction`.

I can look at removing the `Builder get builder` getter from `BuildAction` as well and making `PackageBuilder` not extend builder but think that will be more involved.

Reply:
Ok updated, this turned out to be an easy change.

You still have a normal `BuildStep` in `PackageBuilder.build`, but that would be a larger change to make a `PackageBuildStep`. It might be worth making the change though as I could remove the imports from `src` if we had our own `PackageBuildStepImpl`? But its also a lot of extra code just to clean up that little bit of weirdness :/

Reply:
I think it's fine to have the `BuildStep` argument and src import for now. Since we should have low usage it should be relatively easy to roll a breaking change away from `BuildStep` if we decide we need it. I'm in favor of  moving forward in the mean time.